process/block/interceptedBlocks/interceptedEquivalentProof.go

```diff
@@ -144,12 +160,40 @@ func (iep *interceptedEquivalentProof) CheckValidity() error {
 		return err
 	}
 
+	if !iep.eligibleNodesCache.IsPeerEligible(iep.messageOriginator, iep.proof.GetHeaderShardId(), iep.getEpochForEligibleList()) {
+		return fmt.Errorf("%w, proof originator must be an eligible node", process.ErrInvalidHeaderProof)
+	}
+
+	headerHash := string(iep.proof.GetHeaderHash())
+	iep.km.Lock(headerHash)
+	defer iep.km.Unlock(headerHash)
+
 	ok := iep.proofsPool.HasProof(iep.proof.GetHeaderShardId(), iep.proof.GetHeaderHash())
 	if ok {
 		return common.ErrAlreadyExistingEquivalentProof
 	}
 
-	return iep.headerSigVerifier.VerifyHeaderProof(iep.proof)
+	err = iep.headerSigVerifier.VerifyHeaderProof(iep.proof)
+	if err != nil {
+		return err
+	}
+
+	// also save the proof here in order to complete the flow under mutex lock
+	wasAdded := iep.proofsPool.AddProof(iep.proof)
+	if !wasAdded {
+		// with the current implementation, this should never happen
+		return common.ErrAlreadyExistingEquivalentProof
+	}
+
+	return nil
+}
+func (iep *interceptedEquivalentProof) getEpochForEligibleList() uint32 {
+	epochForEligibleList := iep.proof.GetHeaderEpoch()
+	if iep.proof.GetIsStartOfEpoch() && epochForEligibleList > 0 {
+		return epochForEligibleList - 1
+	}
+
+	return epochForEligibleList
```

Comment:
with the latest merged PR with fields checker we have in common `GetEpochForConsensus` which can be used for this